sip-ri/src/main/java/gov/nist/javax/sip/header/extensions/TargetDialog.java

```diff
@@ -0,0 +1,201 @@
+import gov.nist.javax.sip.header.*;
+
+/**
+ * SIP Target-Dialog header implementation.
+ */
+public class TargetDialog extends ParametersHeader implements ExtensionHeader, TargetDialogHeader {
+
+	  private static final long serialVersionUID = 1L;
+
+	  public CallIdentifier callId;
+
+	  public static final String NAME = SIPHeaderNames.TARGET_DIALOG;
+
+	  private String localTag;
+	  private String remoteTag;
+
+  /**
+   * Default constructor.
+   */
+  public TargetDialog() {
+    super(NAME);
+  }
+
+  /**
+   * Constructor with CallIdentifier.
+   *
+   * @param cid CallIdentifier to set
+   */
+  public TargetDialog(CallIdentifier cid) {
+    super(NAME);
+    callId = cid;
+  }
+
+  /**
+   * Sets the Call-Id of the Target-Dialog header.
+   *
+   * @param callId the Call-Id value to set
+   * @throws ParseException if the provided callId cannot be parsed
+   */
+  public void setCallId(String callId) throws ParseException {
+    try {
+      // Instantiating a new CallIdentifier object using the provided callId
+      this.callId = new CallIdentifier(callId);
+    } catch (Exception e) {
+      // If an exception occurs during instantiation, throw a ParseException
+      throw new ParseException(e.getMessage(), 0);
+    }
+  }
+
+  /**
+   * Gets the Call-Id value of the Target-Dialog header.
+   *
+   * @return the Call-Id value
+   */
+  public String getCallId() {
+    return callId != null ? callId.encode() : null;
+  }
+
+  // New methods for local and remote tags
+
+  /**
+   * Gets the local tag of the Target-Dialog header.
+   *
+   * @return the local tag value
+   */
+  public String getLocalTag() {
+    return localTag;
+  }
+
+  /**
+   * Sets the local tag of the Target-Dialog header.
+   *
+   * @param localTag the local tag value to set
+   * @throws ParseException if the provided localTag is invalid
+   */
+  public void setLocalTag(String localTag) throws ParseException {
+    if (localTag == null || localTag.trim().isEmpty()) {
+      throw new ParseException("Local tag cannot be null or empty", 0);
+    }
+    this.localTag = localTag.trim();
+  }
+
+  /**
+   * Gets the remote tag of the Target-Dialog header.
+   *
+   * @return the remote tag value
+   */
+  public String getRemoteTag() {
+    return remoteTag;
+  }
+
+  /**
+   * Sets the remote tag of the Target-Dialog header.
+   *
+   * @param remoteTag the remote tag value to set
+   * @throws ParseException if the provided remoteTag is invalid
+   */
+  public void setRemoteTag(String remoteTag) throws ParseException {
+    if (remoteTag == null || remoteTag.trim().isEmpty()) {
+      throw new ParseException("Remote tag cannot be null or empty", 0);
+    }
+    this.remoteTag = remoteTag.trim();
+  }
+
+  @Override
+  public StringBuilder encodeBody(StringBuilder retval) {
+    if (callId != null) {
+      retval.append(callId.encode());
+      if (localTag != null) {
+        retval.append(";"); // Use semicolon directly
```

Comment:
as opposed to my previous comment, you completely remove the from-tag while I think it should have been replaced with local-tag. Right now calling getLocalTag would return local-tag=value forcing the application to parse it to remove local-tag=. I think getLocalTag should just return the local tag value

Reply:
got you, refactored it